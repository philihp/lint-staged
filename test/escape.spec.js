/* eslint-disable no-useless-escape */

const { _escapeArg, escapeArg, escapeUnixShellArg, escapeWinCmdArg } = jest.requireActual(
  '../lib/escape'
)

const TEST_STRING = 'file.js'
const UNIX_ESCAPED = escapeUnixShellArg(TEST_STRING)
const WINCMD_ESCAPED = escapeWinCmdArg(TEST_STRING)

describe('escapeArg', () => {
  it('should use current platform by default', () => {
    const isWin = process.platform === 'win32'
    expect(escapeArg(TEST_STRING)).toEqual(isWin ? WINCMD_ESCAPED : UNIX_ESCAPED)
  })

  test.each([
    ['darwin', UNIX_ESCAPED],
    ['linux', UNIX_ESCAPED],
    ['win32', WINCMD_ESCAPED],
    ['cygwin', UNIX_ESCAPED],
  ])('should export platform-specific method for `%s`', async (platform, escaped) => {
    expect(_escapeArg(TEST_STRING, platform)).toEqual(escaped)
  })
})

/**
 * Tricky things to escape
 * @see https://qntm.org/cmd
 */
const TEST_STRINGS = `yes
no
child.exe
argument 1
Hello, world
Hello"world
\some\path with\spaces
C:\Program Files\
she said, "you had me at hello"
arg;,ument"2
\some\directory with\spaces\
"
\
\\
\\\
\\\\
\\\\\
"\
"\T
"\\T
!1
!A
"!\/'"
"Jeff's!"
$PATH
%PATH%
&
<>|&^
()%!^"<>&|
>\\.\nul
malicious argument"&whoami
*@$$A$@#?-_`.split('\n')

describe('escapeUnixShellArg', () => {
  test.each(TEST_STRINGS.map((string) => [string]))('handles `%s`', (string) => {
    expect(escapeUnixShellArg(string)).toMatchSnapshot()
  })
})

describe('escapeWinCmdArg', () => {
  test.each(TEST_STRINGS.map((string) => [string]))('handles `%s`', (string) => {
    expect(escapeWinCmdArg(string)).toMatchSnapshot()
  })
})
