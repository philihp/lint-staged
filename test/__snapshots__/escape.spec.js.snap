// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`escapeUnixShellArg handles \`!1\` 1`] = `"\\\\!1"`;

exports[`escapeUnixShellArg handles \`!A\` 1`] = `"\\\\!A"`;

exports[`escapeUnixShellArg handles \`"!/'"\` 1`] = `"\\\\\\"\\\\!\\\\/\\\\'\\\\\\""`;

exports[`escapeUnixShellArg handles \`"\\T\` 1`] = `"\\\\\\"\\\\\\\\T"`;

exports[`escapeUnixShellArg handles \`"Jeff's!"\` 1`] = `"\\\\\\"Jeff\\\\'s\\\\!\\\\\\""`;

exports[`escapeUnixShellArg handles \`$PATH\` 1`] = `"\\\\$PATH"`;

exports[`escapeUnixShellArg handles \`%PATH%\` 1`] = `"\\\\%PATH\\\\%"`;

exports[`escapeUnixShellArg handles \`&\` 1`] = `"\\\\&"`;

exports[`escapeUnixShellArg handles \`()%!^"<>&|\` 1`] = `"\\\\(\\\\)\\\\%\\\\!\\\\^\\\\\\"\\\\<\\\\>\\\\&\\\\|"`;

exports[`escapeUnixShellArg handles \`*@$$A$@#?-_\` 1`] = `"\\\\*\\\\@\\\\$\\\\$A\\\\$\\\\@\\\\#\\\\?\\\\-_"`;

exports[`escapeUnixShellArg handles \`<>|&^\` 1`] = `"\\\\<\\\\>\\\\|\\\\&\\\\^"`;

exports[`escapeUnixShellArg handles \`>\\.\` 1`] = `"\\\\>\\\\\\\\\\\\."`;

exports[`escapeUnixShellArg handles \`\\\\""T\` 1`] = `"\\\\\\\\\\\\\\\\\\\\\\"\\\\\\"T"`;

exports[`escapeUnixShellArg handles \`\\\\\\\` 1`] = `"\\\\\\\\\\\\\\\\\\\\\\\\"`;

exports[`escapeUnixShellArg handles \`\\\` 1`] = `"\\\\\\\\"`;

exports[`escapeUnixShellArg handles \`C:Program Filesshe said, "you had me at hello"\` 1`] = `"C\\\\:Program\\\\\\\\ Filesshe\\\\\\\\ said\\\\,\\\\\\\\ \\\\\\"you\\\\\\\\ had\\\\\\\\ me\\\\\\\\ at\\\\\\\\ hello\\\\\\""`;

exports[`escapeUnixShellArg handles \`Hello"world\` 1`] = `"Hello\\\\\\"world"`;

exports[`escapeUnixShellArg handles \`Hello, world\` 1`] = `"Hello\\\\,\\\\\\\\ world"`;

exports[`escapeUnixShellArg handles \`arg;,ument"2\` 1`] = `"arg\\\\;\\\\,ument\\\\\\"2"`;

exports[`escapeUnixShellArg handles \`argument 1\` 1`] = `"argument\\\\\\\\ 1"`;

exports[`escapeUnixShellArg handles \`child.exe\` 1`] = `"child\\\\.exe"`;

exports[`escapeUnixShellArg handles \`malicious argument"&whoami\` 1`] = `"malicious\\\\\\\\ argument\\\\\\"\\\\&whoami"`;

exports[`escapeUnixShellArg handles \`no\` 1`] = `"no"`;

exports[`escapeUnixShellArg handles \`somedirectory withspaces"\` 1`] = `"somedirectory\\\\\\\\ withspaces\\\\\\""`;

exports[`escapeUnixShellArg handles \`somepath withspaces\` 1`] = `"somepath\\\\\\\\ withspaces"`;

exports[`escapeUnixShellArg handles \`ul\` 1`] = `"ul"`;

exports[`escapeUnixShellArg handles \`yes\` 1`] = `"yes"`;

exports[`escapeWinCmdArg handles \`!1\` 1`] = `"!1"`;

exports[`escapeWinCmdArg handles \`!A\` 1`] = `"!A"`;

exports[`escapeWinCmdArg handles \`"!/'"\` 1`] = `"\\"!/'\\""`;

exports[`escapeWinCmdArg handles \`"\\T\` 1`] = `"\\"\\\\T"`;

exports[`escapeWinCmdArg handles \`"Jeff's!"\` 1`] = `"\\"Jeff's!\\""`;

exports[`escapeWinCmdArg handles \`$PATH\` 1`] = `"$PATH"`;

exports[`escapeWinCmdArg handles \`%PATH%\` 1`] = `"%PATH%"`;

exports[`escapeWinCmdArg handles \`&\` 1`] = `"&"`;

exports[`escapeWinCmdArg handles \`()%!^"<>&|\` 1`] = `"()%!^\\"<>&|"`;

exports[`escapeWinCmdArg handles \`*@$$A$@#?-_\` 1`] = `"*@$$A$@#?-_"`;

exports[`escapeWinCmdArg handles \`<>|&^\` 1`] = `"<>|&^"`;

exports[`escapeWinCmdArg handles \`>\\.\` 1`] = `">\\\\."`;

exports[`escapeWinCmdArg handles \`\\\\""T\` 1`] = `"\\\\\\\\\\"\\"T"`;

exports[`escapeWinCmdArg handles \`\\\\\\\` 1`] = `"\\\\\\\\\\\\"`;

exports[`escapeWinCmdArg handles \`\\\` 1`] = `"\\\\"`;

exports[`escapeWinCmdArg handles \`C:Program Filesshe said, "you had me at hello"\` 1`] = `"C:Program Filesshe said, \\"you had me at hello\\""`;

exports[`escapeWinCmdArg handles \`Hello"world\` 1`] = `"Hello\\"world"`;

exports[`escapeWinCmdArg handles \`Hello, world\` 1`] = `"Hello, world"`;

exports[`escapeWinCmdArg handles \`arg;,ument"2\` 1`] = `"arg;,ument\\"2"`;

exports[`escapeWinCmdArg handles \`argument 1\` 1`] = `"argument 1"`;

exports[`escapeWinCmdArg handles \`child.exe\` 1`] = `"child.exe"`;

exports[`escapeWinCmdArg handles \`malicious argument"&whoami\` 1`] = `"malicious argument\\"&whoami"`;

exports[`escapeWinCmdArg handles \`no\` 1`] = `"no"`;

exports[`escapeWinCmdArg handles \`somedirectory withspaces"\` 1`] = `"somedirectory withspaces\\""`;

exports[`escapeWinCmdArg handles \`somepath withspaces\` 1`] = `"somepath withspaces"`;

exports[`escapeWinCmdArg handles \`ul\` 1`] = `"ul"`;

exports[`escapeWinCmdArg handles \`yes\` 1`] = `"yes"`;
